### ISSUE TITLE
Buffer size dependency in UTF8JsonGenerator writeRaw

### ISSUE THREAD

**Issue by umcodemonkey** [FasterXML/jackson-core#354]
UT8JsonGenerator improperly depends on the relative sizes of `_outputBuffer` and `_charBuffer`; if the former is more than three times the latter, `writeRaw` will attempt to write too many characters to the character buffer and produce an `ArrayIndexOutOfBoundsException`. Normally this does not occur, as the default sizes for the byte buffer and character buffer are 8000 and 4000 respectively. But the CBOR module asks for a 16k buffer, when can then be recycled to the UTF8JsonGenerator, triggering the error. The test below can reproduce it,.

```java
package org.fizmo;

import com.fasterxml.jackson.annotation.JsonRawValue;
import com.fasterxml.jackson.annotation.JsonValue;
import com.fasterxml.jackson.core.JsonFactory;
import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.dataformat.cbor.CBORFactory;
import com.fasterxml.jackson.dataformat.cbor.CBORGenerator;
import org.junit.Assert;
import org.junit.Test;

public class JacksonCborTest {

    public static class RawBean {
        public String value;

        public RawBean(String value) {
            this.value = value;
        }

        @JsonValue
        @JsonRawValue
        public String getValue() {
            return value;
        }
    }

    @Test
    public void testCase() {
        String data = "{\"x\":\"" + generate(5000) + "\"}";
        RawBean bean = new RawBean(data);
        ObjectMapper mapper = new ObjectMapper();
        mapper.findAndRegisterModules();
        JsonFactory factory = new JsonFactory(mapper);
        CBORFactory cborFactory = new CBORFactory(mapper);

        try {
            try (CBORGenerator generator = cborFactory.createGenerator(System.out)) {
                //generator.writeObject(1);
            }
            JsonGenerator generator = factory.createGenerator(System.out);
            generator.writeObject(bean);
        } catch (Exception e) {
            Assert.fail(e.getMessage());
        }

    }

    private static final String ALPHA_NUMERIC_STRING = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";

    private String generate(int count) {
        StringBuilder builder = new StringBuilder();
        while (count-- != 0) {
            int character = (int)(Math.random()*ALPHA_NUMERIC_STRING.length());
            builder.append(ALPHA_NUMERIC_STRING.charAt(character));
        }
        return builder.toString();
    }
}
```

**Comment by cowtowncoder** [comment]
Will actually move under CBOR just because tests that use CBOR module must be done there... even if fix itself is here.

Well, actually, I may keep both, and maybe re-title this one.

**Comment by cowtowncoder** [comment]
Fixed. Didn't look like `WriterBasedJsonGenerator` has same issue, nor other methods, unless I missed something (and if I did please let me know).

**Comment by cowtowncoder** [comment]
@umcodemonkey should be fixed now, for 2.8.7 (hoping to finalize relatively soon, within week, minor Avro things to complete).

**Comment by umcodemonkey** [comment]
Thanks for the quick turnaround!
